### PR TITLE
#297 tidy up dgraph data

### LINKED ISSUE #297: Tidy up dgraph data
Placeholder issue for tidying up dgraph data and migration scripts.

## Expected Behaviour

- Product data organised by version.
- Product data stored in easily serialised format, e.g. `csv` (mapped to `json`, `rdf`?)
- Migration scripts organised by version, import and export scripts for converting between dgraph internal storage and human-readable data.
- Remove binary files which cannot be easily version controlled, i.e. replace `xlsx` files with `csv`.

## Current Behaviour

- Data not organised clearly by version.
- Inconsistencies between how data/migration files are named/stored.
- Product data stored as `xlsx` and `xlsm` binary files.

## Possible Solution

See above.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #297.

## Description

Updates dgraph data scripts.

### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [ ] Documentation OK.

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [ ] Tests OK.

### CODE REVIEW

**Review by katherine-sussol — COMMENTED** [review]
✨  Thanks for cleaning this up - much shinier and better structured now (typical 🐈 s - leaving a bit of a mess behind them)

It's mostly looking good - I haven't had a chance to go through everything in depth but dropped in a few notes with things I noticed today! Will look more next week.

**Review by mark-prins — COMMENTED** [review]
very impressive!
as noted, it needs some small changes to handle consumables.. and then can we run it and update the server?

File: unified-codes/tools/scripts/dgraph/setup.sh [hunk 1]
```diff
@@ -4,9 +4,9 @@ echo " * Deleting existing graph data..."
 bash ./clean.sh
 
 echo " * Initialising demo graph schema..."
-bash ./alter.sh ./data/schema.gql
+bash ./alter.sh ./data/v2/schema.gql
 
 echo " * Populating demo graph data..."
-bash ./mutate.sh ./data/import.json
+npx ts-node ./migration/v2/import.js ./data/v2/products.csv
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/setup.sh]: 6d274df3b858dfa2c49e32f68fdd5e9556d14316.

**Review by katherine-sussol — COMMENTED** [review]
Maybe for a later PR (separate issues?) - looks like we'll also need to revisit synonyms and properties at some stage. 

Unfortunately, seems we're a bit blocked on both atm 😿  
- Synonyms by Dgraph (which doesn't allow `@*` to search all languages in filter 🙀 )
- Properties by not having a spreadsheet with them filled in

Aside from that, once the consumables spawn some children I think everything else looks great 😎 🎉

File: unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            combinations.forEach(uc => {
+                 if (uc2 && uc) {
+                    if (!(this.tree[uc2].combines.map(sibling => sibling.code).includes(uc))) {
+                        this.tree[uc2].combines.push({ code: uc });
+                    }
+                }
+            });
+
+            // If not linked, link parent product node to child.
+            if (uc2 && uc3) {
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts]: Eep. 8aa5a305c70e87302e0efbaebf9dc2fbaaa2dc4b. Should do the job, but might be a few edge cases still missing?

File: unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    
+            while (stack.length > 0 && !hasCycle) {
+                const entity = stack.pop();
+    
+                if (!visited[entity.code]) {
+                    visited[entity.code] = true;
+                }
+    
+                if (entity.children) {
+                    for (let i = 0; i < entity.children.length && !hasCycle; i++) {
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts]: 5c2064da7484f9b4c169cba39c3949d311af65a2 should do the job.

File: unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        autoClose?: boolean;
+        emitClose?: boolean;
+        start?: number;
+        end?: number;
+        highWaterMark?: number;
+    }) {
+        super(path, options);
+    }
+
+    public validateTree(): boolean {
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts]: 5c2064da7484f9b4c169cba39c3949d311af65a2 🤓 .

**Comment by wlthomson** [comment]
@katherine-sussol @mark-prins This should be backwards compatible with mSupply now!

File: unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+                    //uc9
+                } = row;
+
+                // If row include strength code...
+                if (uc6) {
+                    // create strength node.
+                    if (!(uc6 in this.graph)) {
+                        const code = uc6;
+                        const name = strength;
+                        const type = 'Strength';
```
Comment (katherine-sussol) [inline unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts:218]: `Strength`, or `DoseStrength`? (as per DgraphDataSource)?

**Review by katherine-sussol — CHANGES_REQUESTED** [review]
I'm sorry to be a huge pain 😿  - just gave it a test and noticed some nodes where I think they shouldn't be, haven't pinpointed the root cause yet but here's a screenshot:
Screen Shot 2020-12-01 at 2 54 16 PM

File: unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts (hunk 5 above)
Comment (wlthomson) [reply unified-codes/tools/scripts/dgraph/migration/v2/DataParser.ts]: Oops. d0a7d7637e6c1b3b4b2e80456d0152b3ca3ad752.

**Comment by wlthomson** [comment]
> I'm sorry to be a huge pain 😿 - just gave it a test and noticed some nodes where I think they shouldn't be, haven't pinpointed the root cause yet but here's a screenshot:
> Screen Shot 2020-12-01 at 2 54 16 PM

Hmm, looks like an error in the data... 

Lines `1075-1076`:

```
Verapamil,,,Oral,Tablet,Immediate-release,40mg,,,,,933f3f00,6331f543,ca50a600,dec4d100,77fcbb00,673c04bf,,,,,11170,,
Verapamil,,,Oral,Tablet,Immediate-release,80mg,,,,,933f3f00,6331f543,ca50a600,dec4d100,77fcbb00,673f04bf,,,,,11170,,
```

Note that `77fcbb00` is the code for the consumables category (it looks like the code has accidentally been reused for the `FormQualifier` entity for immediate-release Verapamil). The cycle detection is only on a per-category basis, so this slipped through the net. I wonder if I should add a placeholder root node above the categories to catch these sorts of things?

**Comment by wlthomson** [comment]
@katherine-sussol Have added some code to detect duplicated for different entity types...

```
WARNING: Detected duplicate code 933f3f00!
WARNING: Detected duplicate code 888e7a00!
WARNING: Detected duplicate code 77fcbb00!
WARNING: Detected duplicate code 77fcbb00!
```

Think this explains what you're seeing!

**Review by katherine-sussol — APPROVED** [review]
Dog LGTM Keyboard
Thanks for the link to these GIFs, I will now use these for PRs.

**Pull request merged**